docs/crates/vm.md

```diff
@@ -1,109 +1,222 @@
+    where
+        Ctx: ExecutionCtxTrait;
+}
+```
+
+where `ExecuteFunc<F, Ctx>` is a function pointer that contains the instruction execution logic.
+
+```rust
+pub type ExecuteFunc<F, CTX> =
+    unsafe fn(pre_compute: &[u8], exec_state: &mut VmExecState<F, GuestMemory, CTX>);
+```
+
+Each executor pre-computes instruction-specific data during a preprocessing step and returns function pointers for direct instruction execution.
+
+#### Metered Execution
+
+Metered execution tracks the trace heights for each chip along with normal execution. This mode divides the execution into segments, where each segment consists of an instruction range and an (over)estimate of the resulting trace heights for each chip in the segment. Segmentation is done based on configurable limits like maximum trace height, maximum trace cells etc.
+
+The `MeteredExecutor<F>` trait defines the interface for metered execution:
+
+```rust
+pub trait MeteredExecutor<F> {
+    fn metered_pre_compute_size(&self) -> usize;
+
+    fn metered_pre_compute<Ctx>(
+        &self,
+        air_idx: usize,
+        pc: u32,
+        inst: &Instruction<F>,
+        data: &mut [u8],
+    ) -> Result<ExecuteFunc<F, Ctx>, StaticProgramError>
+    where
+        Ctx: MeteredExecutionCtxTrait;
+}
+```
+
+The additional `air_idx` parameter is the index of the chip's AIR in the verifying key. This is used for indexing the trace height of the chip in the `trace_heights` array contained in the `Segment` struct.
+
+#### Preflight Execution
+
+Preflight execution creates execution records of the record arena type `RA` which are needed for trace generation. Preflight execution doesn't have a precompute mechanism and uses dynamic dispatch to execute each instruction.
+
+The `PreflightExecutor<F, RA>` trait defines the interface for preflight execution:
 
 ```rust
-pub trait PreflightExecutor<F> {
-    /// Runtime execution of the instruction, if the instruction is owned by the
-    /// current instance. May internally store records of this call for later trace generation.
+pub trait PreflightExecutor<F, RA> {
     fn execute(
-        &mut self,
+        &self,
+        state: VmStateMut<F, TracingMemory, RA>,
         instruction: &Instruction<F>,
-        from_state: ExecutionState<u32>,
-  ) -> Result<ExecutionState<u32>>;
+    ) -> Result<(), ExecutionError>;
 }
 ```
 
-There is a `struct VmOpcode(usize)` to protect the global opcode `usize`, which must be globally unique for each opcode
-supported in a given VM.
+### Interpreter Architecture
+
+The `InterpretedInstance` represents the VM interpreter and handles pure and metered execution modes. More specifically, it:
+
+- Pre-computes instruction-specific data buffers
+- Generates function pointer tables for direct execution
+- Supports optional tail call optimization (TCO) for improved performance
+
+The `PreflightInterpretedInstance` handles preflight execution with:
+
+- Dynamic instruction dispatch
```

Comment:
what's the right term here, enum or switch statement style dispatch? I get what you mean, but dynamic dispatch just means a specific thing in rust